File: LC/[3018] Maximum Number of Removal Queries That Can Be Processed I.java
      /* 
       */
      class Solution {
        public int maximumProcessableQueries(int[] nums, int[] queries) {
            int N = nums.length;//java take array size
            int M = queries.length;
            int[][] dp = new int[1005][1005];//java initialize two dimentional array
            dp[0][N-1]=0;//this is known
    
            //type V DP template
            for(int len=N-1; len>=1; --len){//try each len from large interval to smaller as we know the initial status only @ max interval
                for(int ii=0; ii+len-1<N; ++ii){ //each staring idx of the interval
                    int jj = ii+len-1;//ending index
                    //current (smaller) state is from previous (bigger) state
                    /* case I: can we reach current state by eliminate nums[ii] in the interval
                    */
                    if(ii-1>=0){//we can derive current state from ii-1   
                        int kk = dp[ii-1][jj]; 
                        if(kk<M && queries[kk] <= nums[ii-1])
                            dp[ii][jj] = Math.max(dp[ii][jj], dp[ii-1][jj]+1);
                        else
                            dp[ii][jj] = Math.max(dp[ii][jj], dp[ii-1][jj]);
                    }
    
                    // case II: can we reach current state by eliminate nums[jj] in the interval
                    if(jj+1 < N){//we can derive current state from jj+1   
                        int kk = dp[ii][jj+1];//current state??   
                        if(kk<M && queries[kk] <= nums[jj+1])
                            dp[ii][jj] = Math.max(dp[ii][jj], 1+dp[ii][jj+1]);
                        else
                            dp[ii][jj] = Math.max(dp[ii][jj], dp[ii][jj+1]);       
                     }
                }
            }
    
            int ans=0;
            //can we eliminate the last char?
            for(int ii=0; ii<N; ++ii){
                int kk = dp[ii][ii];
                ans = Math.max(ans, kk<M && queries[kk] <= nums[ii] ? kk+1: kk);
            }
            return ans;
        }
    }

File: LC/[3018] Maximum Number of Removal Queries That Can Be Processed I.py
class Solution:
    def maximumProcessableQueries(self, nums: List[int], queries: List[int]) -> int:
        N, M = len(nums), len(queries)

        dp = [[0]*N for _ in range(N)]
        #dp[ii][jj]: the max query we can perform for nums in interavl [ii, jj]

        # initialize the DP
        dp[0][N-1] = 0 # we can perform no query when we have full interval (eliminating nothing)

        # DP is moving from known state to unknown state.  since we know intervanl [0, N-1]
        for ll in range(N, 0, -1):  # from full interval to 1
            for ii in range(N-ll+1): # for each interval start. //ii+len-1 <= N-1 -> ii <= ii-len
                jj=ii+ll-1

                # current state from [ii-1, jj]
                if ii-1 >= 0:
                    kk = dp[ii-1][jj]
                    if kk < M and nums[ii-1] >= queries[kk]:
                        dp[ii][jj] = max(dp[ii][jj], 1+dp[ii-1][jj])
                    else:
                        dp[ii][jj] = max(dp[ii][jj], dp[ii-1][jj])

                # current state from [ii, jj+1]
                if jj+1 < N:
                    kk = dp[ii][jj+1]
                    if kk < M and nums[jj+1] >= queries[kk]:
                        dp[ii][jj] = max(dp[ii][jj], 1+dp[ii][jj+1])
                    else:
                        dp[ii][jj] = max(dp[ii][jj], dp[ii][jj+1])

        ans=0

        for ii in range(N):
            kk = dp[ii][ii]
            if kk < M and nums[ii] >= queries[kk]:
                ans = max(ans, 1+kk)
            else:
                ans = max(ans, kk)
            

        return ans